docs/book/src/developer/providers/contracts/infra-cluster.md

```diff
@@ -0,0 +1,516 @@
+
+To apply the label to CRDs it’s possible to use commonLabels in your `kustomize.yaml` file, usually in `config/crd`:
+
+```yaml
+commonLabels:
+  cluster.x-k8s.io/v1alpha2: v1alpha1
+  cluster.x-k8s.io/v1alpha3: v1alpha2
+  cluster.x-k8s.io/v1beta1: v1beta1
+```
+
+An example of this is in the [Kubeadm Bootstrap provider](https://github.com/kubernetes-sigs/cluster-api/blob/release-1.1/controlplane/kubeadm/config/crd/kustomization.yaml).
+
+### InfraCluster, InfraClusterList resource definition
+
+You MUST define a InfraCluster resource.
+The InfraCluster resource name must have the format produced by `sigs.k8s.io/cluster-api/util/contract.CalculateCRDName(Group, Kind)`.
+
+Note: Cluster API is using such a naming convention to avoid an expensive CRD lookup operation when looking for labels from 
+the CRD definition of the InfraCluster resource.
+
+It is a generally applied convention to use names in the format `${env}Cluster`, where ${env} is a, possibly short, name
+for the environment in question. For example `GCPCluster` is an implementation for the Google Cloud Platform, and `AWSCluster`
+is one for Amazon Web Services.
+
+```go
+//+kubebuilder:object:root=true
+// +kubebuilder:resource:path=fooclusters,shortName=foocl,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+
+// FooCluster is the Schema for fooclusters.
+type FooCluster struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Spec FooClusterSpec `json:"spec,omitempty"`
+    Status FooClusterStatus `json:"status,omitempty"`
+}
+
+type FooClusterSpec struct {
+    // See other rules for more details about mandatory/optional fields in InfraCluster spec.
+    // Other fields SHOULD be added based on the needs of your provider.
+}
+
+type FooClusterStatus struct {
+    // See other rules for more details about mandatory/optional fields in InfraCluster status.
+    // Other fields SHOULD be added based on the needs of your provider.
+}
+```
+
+For each InfraCluster resource, you MUST also add the corresponding list resource.
+The list resource MUST be named as `<InfraCluster>List`.
+
+```go
+//+kubebuilder:object:root=true
+
+// FooClusterList contains a list of fooclusters.
+type FooaClusterList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []FooCluster `json:"items"`
+}
+```
+
+### InfraCluster control plane endpoint
+
+Each Cluster needs a control plane endpoint to sit in front of control plane machines.
+Control plane endpoint can be provided in three ways in Cluster API: by the users, by the control plane provider or
+by the infrastructure provider.
+
+In case you are developing an infrastructure provider which is responsible to provide a control plane endpoint for
+each Cluster, the host and port of the generated control plane endpoint SHOULD surface on `spec.controlPlaneEndpoint`
```

Comment:
```suggestion
each Cluster, the host and port of the generated control plane endpoint MUST surface on `spec.controlPlaneEndpoint`
```

Is this not a must if "you are developing an infrastructure provider which is responsible to provide a control plane endpoint for
each Cluster"

Not sure how it could work otherwise